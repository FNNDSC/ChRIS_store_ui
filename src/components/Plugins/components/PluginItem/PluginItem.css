.plugins-list > .plugin-item-card {
  /* display */
  margin-bottom: 0;
  padding: 0;
  
  /* border */
  border-bottom-color: #ededed;
  border-bottom: 4px solid var(--pf-black-200);
  box-shadow: none;
}

.plugin-item-card-body {
  /* color */
  color: #4d5258;
  color: var(--pf-black-700);

  /* font */
  font-size: 1.4em;
}

.plugin-item-name {
  /* color */
  color: #0088ce;
  color: var(--pf-blue-400);
  
  /* font */
  font-size: 1.5em;
  font-weight: 600;
}

.plugin-item-name:hover, .plugin-item-name:focus { text-decoration: none; }

.plugin-item-title {
  /* font */
  font-size: 1em;
  font-weight: 400;
}

.plugin-item-creation {
  /* color */
  color: #72767b;
  color: var(--pf-black-600);

  /* font */
  font-weight: 300;
}

.plugin-item-author {
  /* color */
  color: #4d5258;
  color: var(--pf-black-700);

  /* border */
  border-bottom-color: #d1d1d1;
  border-bottom: 1px solid var(--pf-black-300);

  /* font */
  font-weight: 300;
}

.plugin-item-author:hover,
.plugin-item-author:focus {
  color: #4d5258;
  color: var(--pf-black-700);
  text-decoration: none;
}

@media (min-width: 768px)  {
  .plugins-list > .plugin-item-card {
    /* display */
    padding: 0 20px;
  }
}