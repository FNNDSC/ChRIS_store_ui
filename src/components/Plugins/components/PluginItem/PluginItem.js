import React from 'react';
import PropTypes from 'prop-types';
import { Card, CardBody } from 'patternfly-react';
import { Link } from 'react-router-dom';
import RelativeDate from '../../../RelativeDate/RelativeDate';
import './PluginItem.css';

const dateIsValid = date => new RelativeDate(date).isValid();

const formatDate = date => new RelativeDate(date).format();

const Plugin = props => (
  <Card className="plugin-item-card">
    <CardBody className="plugin-item-card-body">
      <div>
        <div className="row no-flex">
          <Link
            href={`/plugin/${props.id}`}
            to={`/plugin/${props.id}`}
            className="plugin-item-name"
          >
            {props.name}
          </Link>
          <div className="plugin-item-title">{props.title}</div>
          <div className="plugin-item-creation">
            <Link
              href={`/author/${props.author}`}
              to={`/author/${props.author}`}
              className="plugin-item-author"
            >
              {props.author}
            </Link>
            {dateIsValid(props.creationDate) &&
              ` created ${formatDate(props.creationDate)}`
            }
          </div>
        </div>
      </div>
    </CardBody>
  </Card>
);

Plugin.propTypes = {
  title: PropTypes.string.isRequired,
  id: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  author: PropTypes.string.isRequired,
  creationDate: PropTypes.string.isRequired,
};

export default Plugin;
