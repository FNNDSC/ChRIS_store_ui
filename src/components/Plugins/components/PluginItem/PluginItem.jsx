import React from 'react';
import PropTypes from 'prop-types';
import {
  Badge, Card, CardBody, Split, SplitItem,
} from '@patternfly/react-core';
import { StarIcon } from '@patternfly/react-icons';
import { Link } from 'react-router-dom';
import RelativeDate from '../../../RelativeDate/RelativeDate';
import './PluginItem.css';

const PluginItem = ({
  // Need to do this because the property "creation_date" comes from CUBE
  // eslint-disable-next-line camelcase
  name, authors, title, creation_date, modification_date, category, isFavorite, isLoggedIn, onStarClicked,
}) => {
  function renderStarButton() {
    let className = 'plugin-star';

    if (!isLoggedIn) className += ' disabled';

    if (isFavorite) className += ' favorite';

    return <StarIcon className={className} onClick={onStarClicked} />;
  }

  return (
    <Card className="plugin-item-card">
      <CardBody className="plugin-item-card-body">
        <div>
          <div className="row no-flex">
            <Split>
              <SplitItem isFilled><p style={{ fontSize: '0.9em', fontWeight: 'bold' }}>{name}</p></SplitItem>
              <SplitItem><Badge isRead>{category}</Badge></SplitItem>
            </Split>
            <div className="plugin-item-name">
              <Link
                href={`/plugin/${name}`}
                to={`/plugin/${name}`}
              >
                { title }
              </Link>
              {renderStarButton()}
            </div>
            <Link
              href={`/author/${authors}`}
              to={`/author/${authors}`}
              className="plugin-item-author"
            >
              { authors.length > 1 ?? authors.join(', ') }
            </Link>
            <p style={{ color: 'gray', fontWeight: '600', fontSize: 'small' }}>
              {
                RelativeDate.isValid(modification_date) ?
                  `Updated ${new RelativeDate(modification_date).format()}`
                : 
                  `Created ${new RelativeDate(creation_date).format()}`
              }
            </p>
          </div>
        </div>
      </CardBody>
    </Card>
  );
};

Plugin.propTypes = {
  title: PropTypes.string.isRequired,
  id: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  author: PropTypes.string.isRequired,
  creationDate: PropTypes.string.isRequired,
  isLoggedIn: PropTypes.bool,
  isFavorite: PropTypes.bool,
  onStarClicked: PropTypes.func.isRequired,
};

Plugin.defaultProps = {
  isLoggedIn: false,
  isFavorite: false,
};

export default PluginItem;
