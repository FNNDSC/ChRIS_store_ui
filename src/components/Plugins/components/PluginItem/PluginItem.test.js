import React from 'react';
import { shallow } from 'enzyme';
import PluginItem from './PluginItem';

describe('Plugin', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<PluginItem
      title=""
      id=""
      name=""
      author=""
      creationDate=""
    />);
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render Card component with plugin-item-card class', () => {
    expect(wrapper.find('Card.plugin-item-card')).toHaveLength(1);
  });

  it('should render CardBody component with plugin-item-card-body class', () => {
    expect(wrapper.find('CardBody.plugin-item-card-body')).toHaveLength(1);
  });

  it('should render div with row and no-flex classes', () => {
    expect(wrapper.find('div.row')).toHaveLength(1);
  });

  it('should render plugin-item-title div', () => {
    expect(wrapper.find('div.plugin-item-title')).toHaveLength(1);
  });

  it('should render plugin-item-name div', () => {
    expect(wrapper.find('Link.plugin-item-name')).toHaveLength(1);
  });

  it('should render plugin-item-creation div', () => {
    expect(wrapper.find('div.plugin-item-creation')).toHaveLength(1);
  });

  it('should render plugin-item-author Link inside plugin-item-creation', () => {
    expect(wrapper
      .find('div.plugin-item-creation')
      .find('Link.plugin-item-author'))
      .toHaveLength(1);
  });
});

describe('rendered Plugin', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<PluginItem
      title=""
      id=""
      name=""
      author=""
      creationDate=""
    />);
  });

  it('should render the value of title', () => {
    wrapper.setProps({ title: 'testTitle' });
    expect(wrapper.find('div.plugin-item-title').text()).toEqual('testTitle');
  });

  it('should render the value of name', () => {
    wrapper.setProps({ name: 'testName' });
    expect(wrapper
      .find('Link.plugin-item-name')
      .childAt(0)
      .text())
      .toEqual('testName');
  });

  it('should render the value of author', () => {
    wrapper.setProps({ author: 'testAuthor' });
    expect(wrapper
      .find('Link.plugin-item-author')
      .childAt(0)
      .text())
      .toEqual('testAuthor');
  });

  const getPluginCreationText = () => {
    const pluginCreationDiv = wrapper.find('div.plugin-item-creation');
    const receivedText = pluginCreationDiv.text();
    // remove html elements
    return receivedText.replace(/(<.*>)/g, '');
  };

  it('should not render the value of creationDate if not valid date', () => {
    wrapper.setProps({ creationDate: 'testDate' });
    expect(getPluginCreationText()).toEqual('');
  });

  it('should render the date value as relative date if valid date', () => {
    Date.now = jest.fn(() => 1530814238992);
    wrapper.setProps({ creationDate: '2018-06-19T15:29:11.349272Z' });
    expect(getPluginCreationText()).toEqual(' created 16 days ago');
  });
});
