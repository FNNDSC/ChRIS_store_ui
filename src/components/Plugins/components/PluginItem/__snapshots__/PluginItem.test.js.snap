// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Plugin should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Plugin
    author=""
    creationDate=""
    id=""
    name=""
    title=""
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "accented": false,
      "aggregated": false,
      "aggregatedMini": false,
      "cardRef": null,
      "children": <CardBody
        className="plugin-item-card-body"
>
        <div>
                <div
                        className="row no-flex"
                >
                        <Link
                                className="plugin-item-name"
                                href="/plugin/"
                                replace={false}
                                to="/plugin/"
                        />
                        <div
                                className="plugin-item-title"
                        />
                        <div
                                className="plugin-item-creation"
                        >
                                <Link
                                        className="plugin-item-author"
                                        href="/author/"
                                        replace={false}
                                        to="/author/"
                                />
                        </div>
                </div>
        </div>
</CardBody>,
      "className": "plugin-item-card",
      "matchHeight": false,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <div>
          <div
                    className="row no-flex"
          >
                    <Link
                              className="plugin-item-name"
                              href="/plugin/"
                              replace={false}
                              to="/plugin/"
                    />
                    <div
                              className="plugin-item-title"
                    />
                    <div
                              className="plugin-item-creation"
                    >
                              <Link
                                        className="plugin-item-author"
                                        href="/author/"
                                        replace={false}
                                        to="/author/"
                              />
                    </div>
          </div>
</div>,
        "className": "plugin-item-card-body",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="row no-flex"
>
            <Link
                        className="plugin-item-name"
                        href="/plugin/"
                        replace={false}
                        to="/plugin/"
            />
            <div
                        className="plugin-item-title"
            />
            <div
                        className="plugin-item-creation"
            >
                        <Link
                                    className="plugin-item-author"
                                    href="/author/"
                                    replace={false}
                                    to="/author/"
                        />
            </div>
</div>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Link
                className="plugin-item-name"
                href="/plugin/"
                replace={false}
                to="/plugin/"
/>,
              <div
                className="plugin-item-title"
/>,
              <div
                className="plugin-item-creation"
>
                <Link
                                className="plugin-item-author"
                                href="/author/"
                                replace={false}
                                to="/author/"
                />
</div>,
            ],
            "className": "row no-flex",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "",
                "className": "plugin-item-name",
                "href": "/plugin/",
                "replace": false,
                "to": "/plugin/",
              },
              "ref": null,
              "rendered": "",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "",
                "className": "plugin-item-title",
              },
              "ref": null,
              "rendered": "",
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <Link
                    className="plugin-item-author"
                    href="/author/"
                    replace={false}
                    to="/author/"
/>,
                  false,
                ],
                "className": "plugin-item-creation",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "children": "",
                    "className": "plugin-item-author",
                    "href": "/author/",
                    "replace": false,
                    "to": "/author/",
                  },
                  "ref": null,
                  "rendered": "",
                  "type": [Function],
                },
                false,
              ],
              "type": "div",
            },
          ],
          "type": "div",
        },
        "type": "div",
      },
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "accented": false,
        "aggregated": false,
        "aggregatedMini": false,
        "cardRef": null,
        "children": <CardBody
          className="plugin-item-card-body"
>
          <div>
                    <div
                              className="row no-flex"
                    >
                              <Link
                                        className="plugin-item-name"
                                        href="/plugin/"
                                        replace={false}
                                        to="/plugin/"
                              />
                              <div
                                        className="plugin-item-title"
                              />
                              <div
                                        className="plugin-item-creation"
                              >
                                        <Link
                                                  className="plugin-item-author"
                                                  href="/author/"
                                                  replace={false}
                                                  to="/author/"
                                        />
                              </div>
                    </div>
          </div>
</CardBody>,
        "className": "plugin-item-card",
        "matchHeight": false,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <div>
            <div
                        className="row no-flex"
            >
                        <Link
                                    className="plugin-item-name"
                                    href="/plugin/"
                                    replace={false}
                                    to="/plugin/"
                        />
                        <div
                                    className="plugin-item-title"
                        />
                        <div
                                    className="plugin-item-creation"
                        >
                                    <Link
                                                className="plugin-item-author"
                                                href="/author/"
                                                replace={false}
                                                to="/author/"
                                    />
                        </div>
            </div>
</div>,
          "className": "plugin-item-card-body",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="row no-flex"
>
              <Link
                            className="plugin-item-name"
                            href="/plugin/"
                            replace={false}
                            to="/plugin/"
              />
              <div
                            className="plugin-item-title"
              />
              <div
                            className="plugin-item-creation"
              >
                            <Link
                                          className="plugin-item-author"
                                          href="/author/"
                                          replace={false}
                                          to="/author/"
                            />
              </div>
</div>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <Link
                  className="plugin-item-name"
                  href="/plugin/"
                  replace={false}
                  to="/plugin/"
/>,
                <div
                  className="plugin-item-title"
/>,
                <div
                  className="plugin-item-creation"
>
                  <Link
                                    className="plugin-item-author"
                                    href="/author/"
                                    replace={false}
                                    to="/author/"
                  />
</div>,
              ],
              "className": "row no-flex",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": "",
                  "className": "plugin-item-name",
                  "href": "/plugin/",
                  "replace": false,
                  "to": "/plugin/",
                },
                "ref": null,
                "rendered": "",
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "",
                  "className": "plugin-item-title",
                },
                "ref": null,
                "rendered": "",
                "type": "div",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    <Link
                      className="plugin-item-author"
                      href="/author/"
                      replace={false}
                      to="/author/"
/>,
                    false,
                  ],
                  "className": "plugin-item-creation",
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "class",
                    "props": Object {
                      "children": "",
                      "className": "plugin-item-author",
                      "href": "/author/",
                      "replace": false,
                      "to": "/author/",
                    },
                    "ref": null,
                    "rendered": "",
                    "type": [Function],
                  },
                  false,
                ],
                "type": "div",
              },
            ],
            "type": "div",
          },
          "type": "div",
        },
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
